rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -658,125 +669,226 @@ def _use_default_label_features(self, label_ids: np.ndarray) -> List[FeatureArra
-                example for example in training_data if label_attribute in example.data
+            messages = [
+                example for example in messages if self.label_attribute in example.data
             ]
 
-        if not training_data:
-            # no training data are present to train
+        if not messages:
             return RasaModelData()
 
-        (
-            features_for_examples,
-            sparse_feature_sizes,
-        ) = model_data_utils.featurize_training_examples(
-            training_data,
-            attributes_to_consider,
-            entity_tag_specs=self._entity_tag_specs,
-            featurizers=self.component_config[FEATURIZERS],
-            bilou_tagging=self.component_config[BILOU_FLAG],
-        )
-        attribute_data, _ = model_data_utils.convert_to_data_format(
-            features_for_examples, consider_dialogue_dimension=False
+        # Create RasaModelData
+        model_data = RasaModelData(
+            label_key=self.label_key, label_sub_key=self.label_sub_key,
         )
 
-        model_data = RasaModelData(
-            label_key=self.label_key, label_sub_key=self.label_sub_key
+        # Add features for text labels/entities
+        # (Note that label/entities will only be included if training is False)
+        features, sparse_feature_sizes = self._collect_features(
+            messages=messages, training=training
+        )
+        attribute_data, _ = model_data_utils.convert_to_data_format(
+            features, consider_dialogue_dimension=False
         )
         model_data.add_data(attribute_data)
+
+        # Add sequence length information (only if SEQUENCE sub_key exists here)
         model_data.add_lengths(TEXT, SEQUENCE_LENGTH, TEXT, SEQUENCE)
-        # Current implementation doesn't yet account for updating sparse
+
+        # Add information on size of sparse features
+        # NOTE: Current implementation doesn't yet account for updating sparse
         # feature sizes of label attributes. That's why we remove them.
-        sparse_feature_sizes = self._remove_label_sparse_feature_sizes(
-            sparse_feature_sizes=sparse_feature_sizes, label_attribute=label_attribute
-        )
+        sparse_feature_sizes.pop(self.label_attribute, None)
         model_data.add_sparse_feature_sizes(sparse_feature_sizes)
 
-        self._add_label_features(
-            model_data, training_data, label_attribute, label_id_dict, training
-        )
+        # Add special ID features for the labels and catch edge cases
+        if training and self.label_attribute is not None:
+            label_ids = self._add_label_id_features(model_data, messages, label_id_dict)
+            # in case there are no features for labels at this point, use the
+            # default label features (cf. `_use_default_label_features`)
+            self._fallback_to_default_label_fatures_if_necessary(
+                label_ids=label_ids, model_data=model_data
+            )
```

Comment:
Optional: It was weird initially that `label_ids` was the return value of `self._add_label_id_features` (the name doesn't indicate that it would return something). Also, I think you can collapse `_add_label_id_features` and `_fallback_to_default_label_features_if_necessary` into one function since the other cannot live without the first really. That would also avoid returning `label_ids` from the first function.

Reply:
I kept it separate because I thought it is good to explicitly point out here that there's a fallback happening and I wasn't happy with any name combination I came up with - But I agree that the `label_ids` returned there are not so nice. `label_ids` is also not ideal (to close to the label_id_dict)... maybe `message_label_ids` or so (if I remember correctly 😅 )....